### PR TITLE
[operator] Manage `gardener-resource-manager` for virtual garden cluster

### PR DESCRIPTION
**Pull request by oliver-goetz**
**How to categorize this PR?**

/area dev-productivity delivery ipcei
/kind enhancement

**What this PR does / why we need it**:
With this PR `gardener-operator` deploys a second `gardener-resource-manager` for the virtual garden cluster.
This `virtual-garden-gardener-resource-manager` provides a `token-request` kubeconfig `gardener-operator` can use to access the virtual cluster instead of a static token kubeconfig.
Thus, the static token kubeconfig is now disabled for the virtual garden `kube-apiserver`. 

**Which issue(s) this PR fixes**:
Part of #7016 

**Special notes for your reviewer**:

**Release note**:

```feature operator
`gardener-operator` is now managing the `gardener-resource-manager` instance as part of the virtual garden cluster control plane. It provides a `TokenRequest` API-based kubeconfig for `gardener-operator` to access the virtual garden cluster. The static token kubeconfig is now unconditionally disabled.
```

### CODE REVIEW

File: pkg/operation/botanist/component/resourcemanager/resource_manager.go [hunk 1]
```diff
@@ -1961,6 +1965,22 @@ func (r *resourceManager) SetSecrets(s Secrets) { r.secrets = s }
...
+	return r.values.NamePrefix + v1beta1constants.DeploymentNameGardenerResourceManager
+}
+
+// GetTargetAccessSecretName is the name of the target access secret for the gardener-resource-manager.
+func (r *resourceManager) GetTargetAccessSecretName() string {
+	if r.values.NamePrefix == "" {
+		return secretNameShootAccess
+	}
+	return fmt.Sprintf("%saccess-%s", r.values.NamePrefix, v1beta1constants.DeploymentNameGardenerResourceManager)
+}
```
Comment (rfranzke) [inline pkg/operation/botanist/component/resourcemanager/resource_manager.go]: Do we really need this? This is just about naming, and we'd have to do this at other places as well to be consistent (e.g., KCM). OTOH, we already have many places were the "seed/shoot" names are (coincidently) reused. I would simply keep the "static" shoot access secret name to avoid introducing more inconsistencies.

File: pkg/operation/botanist/resource_manager_test.go [hunk 2]
```diff
@@ -308,7 +314,7 @@ var _ = Describe("ResourceManager", func() {
 		Context("w/ bootstrapping", func() {
```
Comment (rfranzke) [inline pkg/operation/botanist/resource_manager_test.go:332]: These unit tests should now probably be moved to the `shared/resourcemanager_test.go` file, shouldn't they?

File: pkg/operation/botanist/component/resourcemanager/resource_manager.go [hunk 3]
```diff
@@ -1961,6 +1965,22 @@ func (r *resourceManager) SetSecrets(s Secrets) { r.secrets = s }
 // GetValues returns the current configuration values of the deployer.
 func (r *resourceManager) GetValues() Values { return r.values }
 
+// GetManagedResourceName returns the name for the ManagedResource containing resources deployed to the target cluster.
+func (r *resourceManager) GetManagedResourceName() string {
+	if r.values.NamePrefix == "" {
+		return shootManagedResourceName
+	}
+	return r.values.NamePrefix + v1beta1constants.DeploymentNameGardenerResourceManager
+}
```
Comment (rfranzke) [inline pkg/operation/botanist/component/resourcemanager/resource_manager.go]: Nit: You could add a simple unit test for this new exported function.

File: pkg/operator/controller/garden/reconciler.go [hunk 4]
```diff
@@ -272,6 +274,29 @@ func (r *Reconciler) updateStatusOperationSuccess(ctx context.Context, garden *o
...
+		Cluster: clientcmdv1.Cluster{
+			Server:                   namePrefix + v1beta1constants.DeploymentNameKubeAPIServer,
+			CertificateAuthorityData: clusterCABundleSecret.Data[secretsutils.DataKeyCertificateBundle],
+		},
+		AuthInfo: clientcmdv1.AuthInfo{
+			TokenFile: gardenerutils.PathShootToken,
+		},
+	}
+
+	_, err := secretsManager.Generate(ctx, config, secretsmanager.Rotate(secretsmanager.InPlace))
```
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler.go]: This is almost the same like https://github.com/gardener/gardener/blob/722401da9a85fdb9ebddd1b2521d4697b28fdce1/pkg/operation/botanist/secrets.go#L214-L234

Instead of copying the code, can you rather put it into a function (maybe in `utils/gardener/tokenrequest`) and call it from both places?

File: pkg/operator/controller/garden/reconciler_delete.go [hunk 5]
```diff
@@ -132,6 +140,7 @@ func (r *Reconciler) delete(
 			Dependencies: flow.NewTaskIDs(destroyKubeAPIServer),
 		})
 		syncPointVirtualGardenControlPlaneDestroyed = flow.NewTaskIDs(
+			destroyVirtualGardenGardenerResourceManager,
```
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler_delete.go:153]: Not related to this PR, but maybe you can add `destroyKubeAPIServer` here as well. It's implicitly already part of it due to `destroyEtcd`, but specifying it explicitly makes it more understandable, I think.

File: pkg/operator/controller/garden/reconciler_reconcile.go [hunk 6]
```diff
@@ -229,14 +234,33 @@ func (r *Reconciler) reconcile(
 			Fn:           kubeAPIServer.Wait,
 			Dependencies: flow.NewTaskIDs(deployKubeAPIServer),
 		})
+		generateGenericKubeconfigToken = g.Add(flow.Task{
+			Name: "Generating generic kubeconfig token",
+			Fn: flow.TaskFn(func(ctx context.Context) error {
+				return r.generateGenericTokenKubeconfig(ctx, secretsManager)
+			}),
+			Dependencies: flow.NewTaskIDs(syncPointSystemComponents),
+		})
```
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler_reconcile.go]: Shouldn't this secret also be destroyed as part of the deletion flow?

File: pkg/operator/controller/garden/reconciler_reconcile.go [hunk 7]
```diff
@@ -229,14 +234,33 @@ func (r *Reconciler) reconcile(
...
+		deployVirtualGardenGardenerResourceManager = g.Add(flow.Task{
+			Name: "Deploying virtual-garden-gardener-resource-manager",
```
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler_reconcile.go]:
```suggestion
			Name: "Deploying gardener-resource-manager for virtual garden",
```

File: pkg/operator/controller/garden/reconciler_reconcile.go [hunk 8]
```diff
@@ -229,14 +234,33 @@ func (r *Reconciler) reconcile(
...
+			Fn: flow.TaskFn(func(ctx context.Context) error {
+				return r.deployVirtualGardenGardenerResourceManager(ctx, secretsManager, virtualGardenGardenerResourceManager)
+			}),
+			Dependencies: flow.NewTaskIDs(generateGenericKubeconfigToken, waitUntilKubeAPIServerIsReady),
+		})
+		waitUntilVirtualGardenGardenerResourceManagerIsReady = g.Add(flow.Task{
+			Name:         "Waiting until virtual-garden-gardener-resource-manager rolled out",
```
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler_reconcile.go]:
```suggestion
			Name:         "Waiting until gardener-resource-manager for virtual garden rolled out",
```

File: pkg/operator/controller/garden/reconciler_reconcile.go (hunk 6 above)
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler_reconcile.go]: Can we do this earlier? We should probably consider this kubeconfig secret as part of the "system components" / "system resources"?

File: pkg/operator/controller/garden/reconciler_reconcile.go [hunk 9]
```diff
@@ -229,14 +234,33 @@ func (r *Reconciler) reconcile(
...
+		})
 		initializeVirtualClusterClient = g.Add(flow.Task{
 			Name: "Initializing connection to virtual garden cluster",
 			Fn: func(ctx context.Context) error {
 				var err error
 				virtualClusterClient, err = r.initializeVirtualClusterClient(secretsManager)
 				return err
 			},
-			Dependencies: flow.NewTaskIDs(waitUntilKubeAPIServerIsReady),
+			Dependencies: flow.NewTaskIDs(waitUntilVirtualGardenGardenerResourceManagerIsReady),
```
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler_reconcile.go]: Why did you change this as part of this commit?

File: extensions/pkg/controller/controlplane/genericactuator/actuator.go [hunk 10]
```diff
@@ -67,8 +67,8 @@ type ValuesProvider interface {
 // the values provided by the given values provider.
 func NewActuator(
 	providerName string,
-	secretConfigs func(namespace string) []extensionssecretsmanager.SecretConfigWithOptions, shootAccessSecrets func(namespace string) []*gardenerutils.ShootAccessSecret,
-	exposureSecretConfigs func(namespace string) []extensionssecretsmanager.SecretConfigWithOptions, exposureShootAccessSecrets func(namespace string) []*gardenerutils.ShootAccessSecret,
+	secretConfigs func(namespace string) []extensionssecretsmanager.SecretConfigWithOptions, shootAccessSecrets func(namespace string) []*gardenerutils.TargetAccessSecret,
```
Comment (rfranzke) [inline extensions/pkg/controller/controlplane/genericactuator/actuator.go]: As said before, I'd vote to drop this commit. For existing `Shoot`s, the `Secret`s have already `shoot-access` in their names. There is too much effort in making all this generic, and especially too much in making it consistent.

File: pkg/utils/gardener/access.go [hunk 11]
```diff
@@ -0,0 +1,160 @@
+// Copyright 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
```
Comment (rfranzke) [inline pkg/utils/gardener/access.go]: 2023

File: pkg/operation/botanist/component/gardeneraccess/access.go [hunk 12]
```diff
@@ -59,6 +58,8 @@ type gardener struct {
 
 // Values contains configurations for the component.
 type Values struct {
+	// ManagedResourceName is the name of the managed resource for gardeneraccess
+	ManagedResourceName string
```
Comment (rfranzke) [inline pkg/operation/botanist/component/gardeneraccess/access.go]: In the GRM component, we have a `NamePrefix` field and a `GetManagedResourceName()` interface function. Why are you following another approach here?

File: pkg/utils/gardener/tokenrequest/secrets_test.go [hunk 13]
```diff
@@ -82,4 +82,103 @@ var _ = Describe("Secrets", func() {
...
+current-context: foo
+kind: Config
+preferences: {}
+users:
+- name: garden
+  user:
+    token: bar
+`
+		)
+		DescribeTable("IsTokenPopulated",
```
Comment (rfranzke) [inline pkg/utils/gardener/tokenrequest/secrets_test.go]:
```suggestion
		DescribeTable("#IsTokenPopulated",
```

File: pkg/operator/controller/garden/reconciler_delete.go [hunk 14]
```diff
@@ -115,6 +116,10 @@ func (r *Reconciler) delete(
 			Name: "Destroying Kube State Metrics",
 			Fn:   component.OpDestroyAndWait(kubeStateMetrics).Destroy,
 		})
+		destroyVirtualGardenGardenerAccess = g.Add(flow.Task{
+			Name: "Destroying Gardener virtual garden access resources",
+			Fn:   virtualGardenGardenerAccess.Destroy,
+		})
 		destroyVirtualGardenGardenerResourceManager = g.Add(flow.Task{
```
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler_delete.go:124]: Don't we have to wait for the `destroyVirtualGardenGardenerAccess` task before starting to destroy the GRM for the virtual cluster? Otherwise, there could be racy behaviour where this VG-GRM cannot delete the `ManagedResource` for the gardener-access?

File: pkg/operator/controller/garden/reconciler_reconcile.go [hunk 15]
```diff
@@ -253,14 +255,29 @@ func (r *Reconciler) reconcile(
 			Fn:           virtualGardenGardenerResourceManager.Wait,
 			Dependencies: flow.NewTaskIDs(deployVirtualGardenGardenerResourceManager),
 		})
+		deployVirtualGardenGardenerAccess = g.Add(flow.Task{
+			Name:         "Deploying Gardener virtual garden access resources",
```
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler_reconcile.go]:
```suggestion
			Name:         "Deploying resources for gardener-operator access to virtual garden",
```

File: pkg/operator/controller/garden/reconciler_reconcile.go [hunk 16]
```diff
@@ -288,6 +305,24 @@ func (r *Reconciler) reconcile(
...
+		})
+		_ = g.Add(flow.Task{
+			Name: "Deleting old ServiceAccount secrets after rotation of signing key",
+			Fn: flow.TaskFn(func(ctx context.Context) error {
+				return secretsrotation.DeleteOldServiceAccountSecrets(ctx, log, virtualClusterClient, garden.Status.Credentials.Rotation.ServiceAccountKey.LastInitiationFinishedTime.Time)
+			}).
+				RetryUntilTimeout(30*time.Second, 10*time.Minute).
+				DoIf(helper.GetServiceAccountKeyRotationPhase(garden.Status.Credentials) == gardencorev1beta1.RotationCompleting),
+			Dependencies: flow.NewTaskIDs(initializeVirtualClusterClient),
+		})
```
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler_reconcile.go]: Why did you add this? Can you explain how it is related to this PR? There is no `kube-controller-manager` yet which could take over the needed actions.

File: pkg/operator/controller/garden/reconciler_reconcile.go [hunk 17]
```diff
@@ -253,14 +255,29 @@ func (r *Reconciler) reconcile(
...
-			Fn: func(ctx context.Context) error {
+			Fn: flow.TaskFn(func(ctx context.Context) error {
 				var err error
-				virtualClusterClient, err = r.initializeVirtualClusterClient(secretsManager)
+				virtualClusterClient, err = r.initializeVirtualClusterClient(ctx)
 				return err
-			},
-			Dependencies: flow.NewTaskIDs(waitUntilVirtualGardenGardenerResourceManagerIsReady),
+			}).
+				RetryUntilTimeout(5*time.Second, 30*time.Second),
```
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler_reconcile.go]:
```suggestion
				RetryUntilTimeout(time.Second, 30*time.Second),
```

to make this faster :)

**Review by rfranzke — CHANGES_REQUESTED** [review]
Very nice, well-structured PR!

Please also adapt the relevant parts in https://github.com/gardener/gardener/blob/master/docs/concepts/operator.md

File: pkg/operation/botanist/component/resourcemanager/resource_manager.go (hunk 1 above)
Comment (ScheererJ) [reply pkg/operation/botanist/component/resourcemanager/resource_manager.go]: So, you are suggesting that we keep the `shoot-access-...` naming of the secrets for the operator in the `garden` namespace? That is certainly easier in terms of code change. As long as we do not get confused by the names of the resulting artefacts that should be fine.

Comment (rfranzke) [reply pkg/operation/botanist/component/resourcemanager/resource_manager.go]: Yes, that's the suggestion. Obviously, I agree that having something more generic than seed/shoot (like runtime/target) is "more correct"/better, and if there were easy ways to achieve this, I'd be the first to go there, however, as said before, I don't think those exist. Doing this "renaming" consistently is a lot of effort, and the current changes just make it more inconsistent. At least for the time being, I propose to not change this and finish the `gardener-operator` epic for now. Maybe we can work on consistent renaming in the aftermath.

Comment (oliver-goetz) [reply pkg/operation/botanist/component/resourcemanager/resource_manager.go]: If consistent naming is not the first priority, we don't need to introduce `GetManagedResourceName()` method to the resource manager, but keep `shoot-core-gardener-resource-manager` as static name of the managed resource instead.
This would simplify the deployment of resource manager. WDYT?

File: pkg/operator/controller/garden/reconciler_reconcile.go (hunk 16 above)
Comment (oliver-goetz) [reply pkg/operator/controller/garden/reconciler_reconcile.go]: This was an open question we had. However, we forgot to clarify it 😅

File: pkg/operator/controller/garden/reconciler_reconcile.go (hunk 6 above)
Comment (oliver-goetz) [reply pkg/operator/controller/garden/reconciler_reconcile.go]: Yep, this looks cleaner.
We looked at `InitializeSecretsManagement` in `gardenlet` where the generic token kubeconfig is neither be deleted. However, we keep the namespace in the operator case, so it is better to delete it here.

File: pkg/operation/botanist/component/resourcemanager/resource_manager.go (hunk 1 above)
Comment (oliver-goetz) [reply pkg/operation/botanist/component/resourcemanager/resource_manager.go]: as discussed, we keep the static `shoot-...` names for the sake of simplicity

File: pkg/operator/controller/garden/reconciler_reconcile.go (hunk 16 above)
Comment (rfranzke) [reply pkg/operator/controller/garden/reconciler_reconcile.go]: OK, then let's remove it for now

File: pkg/operation/botanist/resource_manager_test.go (hunk 2 above)
Comment (oliver-goetz) [reply pkg/operation/botanist/resource_manager_test.go:332]: That's complicated 😅 
There are lots of tests which belong into the shared component. However, `botanist` package is also used many times and many contexts include the hibernation status of the shoot.

Comment (rfranzke) [reply pkg/operation/botanist/resource_manager_test.go:332]: ?

Comment (oliver-goetz) [reply pkg/operation/botanist/resource_manager_test.go:332]: The tests of the first `Context()` are all using [the botanist package](https://github.com/gardener/gardener/blob/0236302440abe8770dfef78436b1a7acaec58d8d/pkg/operation/botanist/resource_manager_test.go#L416-L454)
I thought we should not import `botanist` from `botanist/component/shared`

Comment (rfranzke) [reply pkg/operation/botanist/resource_manager_test.go:332]: Yes, those tests are specific to the `botanist` package and should remain there. I was just talking about the other test cases that cover the functionality that was now moved to the `shared` package. Keeping all of it in the `botanist` package works, but it also introduces some technical debts...

Comment (oliver-goetz) [reply pkg/operation/botanist/resource_manager_test.go:332]: I created tests in `pkg/operation/botanist/component/shared`.
The old tests are not deleted yet. These tests implicitly test [Botanist.determineControllerReplicas](https://github.com/gardener/gardener/blob/af5b7d40315e80a4d36917814f8046229ba8a543/pkg/operation/botanist/controlplane.go#L40-L68) which is tested nowhere else and looks worth to be tested. When there are dedicated test most of the tests could be dropped.

File: pkg/operation/botanist/component/gardeneraccess/access.go [hunk 18]
```diff
@@ -110,6 +111,15 @@ func (g *gardener) Deploy(ctx context.Context) error {
 }
 
 func (g *gardener) Destroy(ctx context.Context) error {
+	var shootAccessSecretNames = []string{v1beta1constants.SecretNameGardener, v1beta1constants.SecretNameGardenerInternal}
+
+	for _, v := range shootAccessSecretNames {
```
Comment (rfranzke) [inline pkg/operation/botanist/component/gardeneraccess/access.go]:
```suggestion
	for _, v := range []string{v1beta1constants.SecretNameGardener, v1beta1constants.SecretNameGardenerInternal} {
```

File: pkg/operation/botanist/component/gardeneraccess/access.go [hunk 19]
```diff
@@ -110,6 +111,15 @@ func (g *gardener) Deploy(ctx context.Context) error {
...
+		secret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: v, Namespace: g.namespace}}
+		if err := client.IgnoreNotFound(g.client.Delete(ctx, secret)); err != nil {
```
Comment (rfranzke) [inline pkg/operation/botanist/component/gardeneraccess/access.go]: This works, but typically we do it like this:
```suggestion
		if err := g.client.Delete(ctx, secret); client.IgnoreNotFound(err) != nil {
```

File: pkg/operation/botanist/component/gardeneraccess/access.go [hunk 20]
```diff
@@ -110,6 +111,15 @@ func (g *gardener) Deploy(ctx context.Context) error {
...
+			return err
```
Comment (rfranzke) [inline pkg/operation/botanist/component/gardeneraccess/access.go]:
```suggestion
			return fmt.Errorf("failed deleting secret %s: %w", client.ObjectKeyFromObject(secret), err)
```

File: pkg/operation/botanist/component/shared/resourcemanager.go [hunk 21]
```diff
@@ -102,3 +114,174 @@ func NewGardenerResourceManager(
...
+	if err != nil {
+		return err
+	}
+
+	if mustBootstrap {
+		bootstrapKubeconfigSecret, err := reconcileGardenerResourceManagerBootstrapKubeconfigSecret(
+			ctx,
+			secretsManager,
+			namespace,
+			getAPIServerAddress)
```
Comment (rfranzke) [inline pkg/operation/botanist/component/shared/resourcemanager.go]:
```suggestion
			getAPIServerAddress,
		)
```

File: pkg/utils/gardener/tokenrequest/tokenrequest_suite_test.go [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+package tokenrequest_test
+
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+func TestSecretsRotation(t *testing.T) {
```
Comment (rfranzke) [inline pkg/utils/gardener/tokenrequest/tokenrequest_suite_test.go]:
```suggestion
func TestTokenRequest(t *testing.T) {
```

File: pkg/operator/controller/garden/components.go [hunk 23]
```diff
@@ -75,6 +81,50 @@ func (r *Reconciler) newGardenerResourceManager(garden *operatorv1alpha1.Garden,
...
+		},
+		WatchedNamespace: pointer.String(r.GardenNamespace),
+	}
+
+	return resourcemanager.New(
+		r.RuntimeClientSet.Client(),
+		r.GardenNamespace,
+		secretsManager,
+		cfg,
+	), nil
```
Comment (rfranzke) [inline pkg/operator/controller/garden/components.go]: This function is still quite similar to the `botanist.DefaultResourceManager`. Perhaps we can reuse the same parts better?

File: pkg/operator/controller/garden/reconciler_delete.go [hunk 24]
```diff
@@ -174,6 +191,13 @@ func (r *Reconciler) delete(
 			Fn:           component.OpWait(gardenerResourceManager).Destroy,
 			Dependencies: flow.NewTaskIDs(ensureNoManagedResourcesExistAnymore),
 		})
+		cleanupGenericTokenKubeconfig = g.Add(flow.Task{
+			Name: "Cleaning up generic token kubeconfig",
+			Fn: flow.TaskFn(func(ctx context.Context) error {
+				return r.cleanupGenericTokenKubeconfig(ctx, secretsManager)
+			}),
+			Dependencies: flow.NewTaskIDs(destroyGardenerResourceManager),
+		})
```
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler_delete.go]: Why do we need to wait for this task? Can't we destroy it at the same time when the virtual garden GRM is destroyed?

File: test/integration/operator/garden/garden_test.go [hunk 25]
```diff
@@ -425,6 +427,124 @@ var _ = Describe("Garden controller tests", func() {
...
+					Type:               "ResourcesHealthy",
+					Status:             "True",
+					LastUpdateTime:     metav1.NewTime(time.Unix(0, 0)),
+					LastTransitionTime: metav1.NewTime(time.Unix(0, 0)),
+				},
+				{
+					Type:               "ResourcesApplied",
+					Status:             "True",
+					LastUpdateTime:     metav1.NewTime(time.Unix(0, 0)),
+					LastTransitionTime: metav1.NewTime(time.Unix(0, 0)),
```
Comment (rfranzke) [inline test/integration/operator/garden/garden_test.go:487]: Is setting the `LastUpdateTime` and `LastTransitionTime` really needed?

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/operator/controller/garden/components.go (hunk 23 above)
Comment (oliver-goetz) [reply pkg/operator/controller/garden/components.go]: done 😄

File: test/integration/operator/garden/garden_test.go (hunk 25 above)
Comment (oliver-goetz) [reply test/integration/operator/garden/garden_test.go:487]: yes, they are required values. Otherwise the update fails.

**Comment by rfranzke** [comment]
Awesome PR, well done, thanks @oliver-goetz and @ScheererJ 
Let's go 🚀

/lgtm
/approve
